mysys/checksum.c

```diff
@@ -18,6 +18,48 @@
 #include <my_sys.h>
 #include <zlib.h>
 
+/* TODO: remove this once zlib adds inherent support for hardware accelerated
+crc32 for all architectures. */
+typedef unsigned long (*my_crc32_func_t)(unsigned long crc,
+                                         const unsigned char *ptr,
+                                         unsigned int len);
+static my_crc32_func_t my_crc32= crc32;
+
+#if __GNUC__ >= 4 && defined(__x86_64__) && defined(HAVE_CLMUL_INSTRUCTION)
+/*----------------------------- x86_64 ---------------------------------*/
+extern int is_pclmul_enabled();
+extern unsigned long crc32_pclmul(unsigned long crc32,
+                                  const unsigned char *buf, unsigned int len);
+void my_crc32_init()
+{
+  my_crc32= is_pclmul_enabled() ? crc32_pclmul : crc32;
+}
+#elif defined(__GNUC__) && defined(__linux__) && defined(HAVE_ARMV8_CRC)
+/*----------------------------- aarch64 --------------------------------*/
+extern unsigned int crc32_aarch64_available(void);
+extern unsigned long crc32_aarch64(unsigned long crc32,
+                                   const unsigned char *buf, unsigned int len);
+
+/* Ideally all ARM 64 bit processor should support crc32 but if some model
+doesn't support better to find it out through auxillary vector. */
+void my_crc32_init()
+{
+  my_crc32= crc32_aarch64_available() ? crc32_aarch64 : crc32;
+}
+#elif defined(HAVE_CRC32_VPMSUM)
+/*----------------------------- ppc64{,le} ---------------------------------*/
+extern unsigned int crc32ieee_vpmsum(unsigned int crc, const unsigned char *p,
+                                     unsigned long len);
+void my_crc32_init()
+{
+}
+#else
+void my_crc32_init()
+{
+  my_crc32= crc32;
+}
+#endif
```

Comment:
I think that each compilation unit that provides a special implementation must define `my_checksum_init(void)`, overriding the weak global symbol that I would propose to introduce in this compilation unit. In that way, we should be able to avoid any `#ifdef` here.

Reply:
depends on what we feel on exporting the my_crc32_t as explained above.

Reply:
folded